### PR TITLE
rocAL - MCW changes

### PR DESCRIPTION
**Pull request by shobana-mcw**
Add affinity change to all nodes in amd_rpp. 
Add HIP backend support for all nodes in amd_rpp.
Reduce variants  from 4->1. Retaining only batchPD in amd_rpp.
Set affinity to CPU for all nodes in OCL backend.

### CODE REVIEW

File: amd_openvx_extensions/amd_rpp/source/AbsoluteDifferencebatchPD.cpp [hunk 1]
```diff
@@ -22,209 +22,256 @@ THE SOFTWARE.
...
-	Rpp32u *srcBatch_height = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
-	copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	copy_status = vxCopyArrayRange((vx_array)parameters[3], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	for(int i = 0; i < data->nbatchSize; i++){
-		data->srcDimensions[i].width = srcBatch_width[i];
-		data->srcDimensions[i].height = srcBatch_height[i];
-	}
-	if(data->device_type == AGO_TARGET_AFFINITY_GPU) {
+    vx_status status = VX_SUCCESS;
+     size_t arr_size;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/AbsoluteDifferencebatchPD.cpp:49]: please adjust the spacing here

File: amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp [hunk 2]
```diff
@@ -22,201 +22,246 @@ THE SOFTWARE.
...
-	Rpp32u *srcBatch_height = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
-	copy_status = vxCopyArrayRange((vx_array)parameters[1], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	for(int i = 0; i < data->nbatchSize; i++){
-		data->srcDimensions[i].width = srcBatch_width[i];
-		data->srcDimensions[i].height = srcBatch_height[i];
-	}
-	if(data->device_type == AGO_TARGET_AFFINITY_GPU) {
+    vx_status status = VX_SUCCESS;
+     size_t arr_size;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp:47]: adjust spacing Ln#47 and Ln#49

File: amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp [hunk 3]
```diff
@@ -22,201 +22,246 @@ THE SOFTWARE.
...
-	STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[3], VX_IMAGE_ATTRIBUTE_FORMAT, &df_image, sizeof(df_image)));
-	vxReleaseImage(&input);
-	vxReleaseImage(&output);
-	vxReleaseParameter(&output_param);
-	vxReleaseParameter(&input_param);
-	return status;
+    vx_status status = VX_SUCCESS;
+    vx_enum scalar_type;
+    STATUS_ERROR_CHECK(vxQueryScalar((vx_scalar)parameters[5], VX_SCALAR_TYPE, &scalar_type, sizeof(scalar_type)));
+     if(scalar_type != VX_TYPE_UINT32) return ERRMSG(VX_ERROR_INVALID_TYPE, "validate: Paramter: #5 type=%d (must be size)\n", scalar_type);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp:86]: correct spacing here

File: amd_openvx_extensions/amd_rpp/source/ContrastbatchPD.cpp [hunk 4]
```diff
@@ -22,225 +22,250 @@ THE SOFTWARE.
...
-	Rpp32u *srcBatch_height = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
-	copy_status = vxCopyArrayRange((vx_array)parameters[1], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	for(int i = 0; i < data->nbatchSize; i++){
-		data->srcDimensions[i].width = srcBatch_width[i];
-		data->srcDimensions[i].height = srcBatch_height[i];
-	}
-	if(data->device_type == AGO_TARGET_AFFINITY_GPU) {
+    vx_status status = VX_SUCCESS;
+     size_t arr_size;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/ContrastbatchPD.cpp:48]: please correct spacing here

File: amd_openvx_extensions/amd_rpp/source/CropMirrorNormalizePD.cpp [hunk 5]
```diff
@@ -22,267 +22,292 @@ THE SOFTWARE.
...
-	STATUS_ERROR_CHECK(vxQueryParameter(input_param, VX_PARAMETER_ATTRIBUTE_REF, &input, sizeof(vx_image)));
-	STATUS_ERROR_CHECK(vxQueryImage(input, VX_IMAGE_ATTRIBUTE_FORMAT, &df_image, sizeof(df_image))); 
-	if(df_image != VX_DF_IMAGE_U8 && df_image != VX_DF_IMAGE_RGB) 
-	{
-		return ERRMSG(VX_ERROR_INVALID_FORMAT, "validate: CropMirrorNormalizebatchPD: image: #0 format=%4.4s (must be RGB2 or U008)\n", (char *)&df_image);
-	}
+    vx_status status = VX_SUCCESS;
+    vx_enum scalar_type;
+    STATUS_ERROR_CHECK(vxQueryScalar((vx_scalar)parameters[11], VX_SCALAR_TYPE, &scalar_type, sizeof(scalar_type)));
+     if(scalar_type != VX_TYPE_UINT32) return ERRMSG(VX_ERROR_INVALID_TYPE, "validate: Paramter: #11 type=%d (must be size)\n", scalar_type);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/CropMirrorNormalizePD.cpp:101]: please fix all the spacing where it is not correct

File: amd_openvx_extensions/amd_rpp/source/LookUpTablebatchPD.cpp [hunk 6]
```diff
@@ -22,201 +22,246 @@ THE SOFTWARE.
...
-	STATUS_ERROR_CHECK(vxSetMetaFormatAttribute(metas[3], VX_IMAGE_ATTRIBUTE_FORMAT, &df_image, sizeof(df_image)));
-	vxReleaseImage(&input);
-	vxReleaseImage(&output);
-	vxReleaseParameter(&output_param);
-	vxReleaseParameter(&input_param);
-	return status;
+    vx_status status = VX_SUCCESS;
+    vx_enum scalar_type;
+    STATUS_ERROR_CHECK(vxQueryScalar((vx_scalar)parameters[5], VX_SCALAR_TYPE, &scalar_type, sizeof(scalar_type)));
+     if(scalar_type != VX_TYPE_UINT32) return ERRMSG(VX_ERROR_INVALID_TYPE, "validate: Paramter: #5 type=%d (must be size)\n", scalar_type);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/LookUpTablebatchPD.cpp:86]: correct spacing

File: amd_openvx_extensions/amd_rpp/source/MaxbatchPD.cpp [hunk 7]
```diff
@@ -22,171 +22,216 @@ THE SOFTWARE.
...
-	Rpp32u *srcBatch_height = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
-	copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	copy_status = vxCopyArrayRange((vx_array)parameters[3], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-	for(int i = 0; i < data->nbatchSize; i++){
-		data->srcDimensions[i].width = srcBatch_width[i];
-		data->srcDimensions[i].height = srcBatch_height[i];
-	}
-	if(data->device_type == AGO_TARGET_AFFINITY_GPU) {
+    vx_status status = VX_SUCCESS;
+     size_t arr_size;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/MaxbatchPD.cpp:49]: correct spacing here

**Review by rrawther — CHANGES_REQUESTED** [review]
@shobana-mcw : Can you please push the commits for spacing adjustment?

File: amd_openvx_extensions/amd_rpp/source/AccumulatebatchPD.cpp [hunk 8]
```diff
@@ -49,13 +50,15 @@ static vx_status VX_CALLBACK refreshAccumulatebatchPD(vx_node node, const vx_ref
     STATUS_ERROR_CHECK(vxQueryImage((vx_image)parameters[0], VX_IMAGE_HEIGHT, &data->maxSrcDimensions.height, sizeof(data->maxSrcDimensions.height)));
     STATUS_ERROR_CHECK(vxQueryImage((vx_image)parameters[0], VX_IMAGE_WIDTH, &data->maxSrcDimensions.width, sizeof(data->maxSrcDimensions.width)));
     data->maxSrcDimensions.height = data->maxSrcDimensions.height / data->nbatchSize;
-    copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),data->srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    copy_status = vxCopyArrayRange((vx_array)parameters[3], 0, data->nbatchSize, sizeof(Rpp32u),data->srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    for(int i = 0; i < data->nbatchSize; i++){
+    copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u), data->srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/AccumulatebatchPD.cpp:53]: change it to STATUS_ERROR_CHECK

File: amd_openvx_extensions/amd_rpp/source/BilateralFilterbatchPD.cpp [hunk 9]
```diff
@@ -49,19 +50,21 @@ static vx_status VX_CALLBACK refreshBilateralFilterbatchPD(vx_node node, const v
 {
     vx_status status = VX_SUCCESS;
     vx_status copy_status;
-    copy_status = vxCopyArrayRange((vx_array)parameters[4], 0, data->nbatchSize, sizeof(vx_uint32),data->kernelSize, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    copy_status = vxCopyArrayRange((vx_array)parameters[5], 0, data->nbatchSize, sizeof(vx_float64),data->sigmaI, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    copy_status = vxCopyArrayRange((vx_array)parameters[6], 0, data->nbatchSize, sizeof(vx_float64),data->sigmaS, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
+    copy_status = vxCopyArrayRange((vx_array)parameters[4], 0, data->nbatchSize, sizeof(vx_uint32), data->kernelSize, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/BilateralFilterbatchPD.cpp:53]: change to STATUS_ERROR_CHECK

File: amd_openvx_extensions/amd_rpp/source/BitwiseNOTbatchPD.cpp [hunk 10]
```diff
@@ -54,13 +55,15 @@ static vx_status VX_CALLBACK refreshBitwiseNOTbatchPD(vx_node node, const vx_ref
     data->srcDimensions = (RppiSize *)malloc(sizeof(RppiSize) * data->nbatchSize);
     Rpp32u *srcBatch_width = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
     Rpp32u *srcBatch_height = (Rpp32u *)malloc(sizeof(Rpp32u) * data->nbatchSize);
-    copy_status = vxCopyArrayRange((vx_array)parameters[1], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    copy_status = vxCopyArrayRange((vx_array)parameters[2], 0, data->nbatchSize, sizeof(Rpp32u),srcBatch_height, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
-    for(int i = 0; i < data->nbatchSize; i++){
+    copy_status = vxCopyArrayRange((vx_array)parameters[1], 0, data->nbatchSize, sizeof(Rpp32u), srcBatch_width, VX_READ_ONLY, VX_MEMORY_TYPE_HOST);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/BitwiseNOTbatchPD.cpp:58]: better to use STATUS_ERROR_CHECK macro instead, wherever applicable

**Review by rrawther — CHANGES_REQUESTED** [review]
@paveltc : Can you please run rali unit tests on this PR?

**Comment by paveltc** [comment]
@rrawther I ran all of the unit tests on both OCL and HIP backends and did not see any new issues. Should be fine to merge.

**Comment by kiritigowda** [comment]
@shobana-mcw can you take a look at these [issues](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=7749504)

File: amd_openvx_extensions/amd_rpp/source/CropMirrorNormalizePD.cpp (hunk 5 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/CropMirrorNormalizePD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/LookUpTablebatchPD.cpp (hunk 6 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/LookUpTablebatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/MaxbatchPD.cpp (hunk 7 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/MaxbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/AccumulatebatchPD.cpp (hunk 8 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/AccumulatebatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/BitwiseNOTbatchPD.cpp (hunk 10 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/BitwiseNOTbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/BilateralFilterbatchPD.cpp (hunk 9 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/BilateralFilterbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/ContrastbatchPD.cpp (hunk 4 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/ContrastbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp (hunk 3 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp (hunk 2 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/ChannelExtractbatchPD.cpp]: Fixed.

File: amd_openvx_extensions/amd_rpp/source/AbsoluteDifferencebatchPD.cpp (hunk 1 above)
Comment (shobana-mcw) [reply amd_openvx_extensions/amd_rpp/source/AbsoluteDifferencebatchPD.cpp]: Fixed.

File: rocAL/rocAL_pybind/CMakeLists.txt [hunk 11]
```diff
@@ -58,7 +58,7 @@ endif()
 if(${BUILD_RALI_PYBIND})
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
 
-    link_directories(${ROCM_PATH}/rpp/lib ${CMAKE_INSTALL_PREFIX}/lib)
+    link_directories(${ROCM_PATH}/rpp/lib ${ROCM_PATH}/mivisionx/lib)
```
Comment (asalmanp) [inline rocAL/rocAL_pybind/CMakeLists.txt:61]: Please keep the change from the master branch for this file. the CMAKE_INSTALL_PREFIX is correct here and shouldn't be changed to ROCM_PATH}/mivisionx/.

**Review by asalmanp — CHANGES_REQUESTED** [review]

File: rocAL/rocAL_pybind/CMakeLists.txt (hunk 11 above)
Comment (rrawther) [reply rocAL/rocAL_pybind/CMakeLists.txt:61]: @asalmanp: somehow the CMAKE_INSTALL_PREFIX didn't have the right path when I tested this out.
Can you please test rocal_pybind module install by following the steps on the readme?

**Comment by rrawther** [comment]
@paveltc : Can you please run the rali unit-tests with this PR?

**Comment by paveltc** [comment]
@rrawther I ran all of the unit tests on this PR yesterday and they passed. Were any changes made that could affect the unit tests?

File: rocAL/rocAL_pybind/CMakeLists.txt (hunk 11 above)
Comment (shobana-mcw) [reply rocAL/rocAL_pybind/CMakeLists.txt:61]: I also faced similar issue. Only if I use ${ROCM_PATH}/mivisionx/lib its able to fetch the libraries.

Comment (kiritigowda) [reply rocAL/rocAL_pybind/CMakeLists.txt:61]: @shobana-mcw @asalmanp & @rrawther the TOT has the `${ROCM_PATH}/mivisionx/lib` path to build and run tests for rocAL pybind. Issue #573 with support to build pybind will improve the solution. @rrawther is this PR ready to be merged?

**Comment by rrawther** [comment]
@kiritigowda: Yes this is ready to be merged and @paveltc tested the unit-tests

**Pull request merged**